mysql/access_rules_test.go

```diff
@@ -0,0 +1,175 @@
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	sInstance, err := sClient.createTestServiceInstance()
+	if err != nil {
+		t.Fatalf("Error creating Service Instance: %s", err)
+	}
+
+	instanceName := sInstance.ServiceName
+	defer destroyServiceInstance(t, sClient, instanceName)
+
+	createAccessRuleInput := &CreateAccessRuleInput{
+		ServiceInstanceID: instanceName,
+		Description:       _Service_AccessRule_Description,
+		Destination:       _Service_AccessRule_Destination,
+		Ports:             _Service_AccessRule_Ports,
+		Protocol:          _Service_AccessRule_Protocol,
+		RuleName:          _Service_AccessRule_Name,
+		Source:            _Service_AccessRule_Source,
+		Status:            _Service_AccessRule_Status,
+	}
+
+	expected := &AccessRuleInfo{
+		Description: _Service_AccessRule_Description,
+		Destination: _Service_AccessRule_Destination,
+		Ports:       _Service_AccessRule_Ports,
+		Protocol:    _Service_AccessRule_Protocol,
+		RuleName:    _Service_AccessRule_Name,
+		Source:      _Service_AccessRule_Source,
+		Status:      _Service_AccessRule_Status,
+		RuleType:    "USER",
+	}
+
+	if err := aClient.CreateAccessRule(createAccessRuleInput); err != nil {
+		t.Fatalf("Error creating AccessRule: %s", err)
+	}
+
+	// Get Access Rule (Create only returns AccessRule name)
+	getInput := &GetAccessRuleInput{
+		ServiceInstanceID: instanceName,
+	}
+
+	allRulesResult, err := aClient.GetAllAccessRules(getInput)
+	if err != nil {
+		t.Fatalf("Error reading ALL AccessRules : %s", err)
+	}
+
+	if len(allRulesResult.AccessRules) == 0 {
+		t.Fatalf("Error reading ALL accessRules: Expected at least 1 rule. Got %d", len(allRulesResult.AccessRules))
+	}
+
+	// Read Result
+	getInput.Name = _Service_AccessRule_Name
+	ruleResult, err := aClient.GetAccessRule(getInput)
+
+	if err != nil {
+		t.Fatalf("Error reading AccessRule: %s", err)
+	}
+
+	// Test Assertions
+	if diff := pretty.Compare(ruleResult, expected); diff != "" {
+		t.Fatalf("Diff creating AccessRule: (-got, +want):\n%s", diff)
+	}
+
+	// Update Access Rule
+	updateInput := &UpdateAccessRuleInput{
+		ServiceInstanceID: instanceName,
+		Name:              _Service_AccessRule_Name,
+		Status:            "disabled",
+	}
+
+	if _, err := aClient.UpdateAccessRule(updateInput); err != nil {
+		t.Fatalf("Error updating AccessRule: %s", err)
+	}
+
+	// Re-Read Result
+	ruleResult, err = aClient.GetAccessRule(getInput)
+	if err != nil {
+		t.Fatalf("Error reading AccessRule: %s", err)
+	}
+
+	// Change expected to match
+	expected.Status = "disabled"
+
+	// Test Assertions
+	if diff := pretty.Compare(ruleResult, expected); diff != "" {
+		t.Fatalf("Diff creating AccessRule: (-got, +want):\n%s", diff)
+	}
+
+	// We don't explicitly test the delete, because its called
+	// using the defer.
+
+	// Sleep for 3 minutes to prevent failure because the resource is still
+	// locked.
+
+	destroyAccessRule(t, aClient, instanceName, _Service_AccessRule_Name)
+	time.Sleep(3 * time.Minute)
```

Comment:
Are we sleeping because the access rule isn't ready? Maybe we should call WaitForReady after we update and then we won't have to sleep here

Reply:
Another unknown problem here.  The deleteAccessRule already waits for the rule deletion to be complete before returning, so the instance is technically ready. For some unknown reason, if I don't sleep, when the defer deleteServiceInstance kicks in, and the deleteServiceInstance fails.

Reply:
Do you think this should be moved into the access rule logic then? I'm just thinking if we don't move it out of the test, we'll have to add the sleep to the terraform code so we might as well just add it to the delete method of the sdk?

Reply:
Let me do a bit more testing here before I make a decision. I suspect there may be a logic error with the WaitForAccessRuleDeleted. The Wait method exits when the API returns the AccessRule list without the deleted rule. But when I look at the activity logs, i see that the deletion is still in RUNNING state, so that may be blocking the ServiceInstance from being deleted.
When eliminate this possiblity, I'll add in my comments in the code and update you again.

Reply:
I managed to fix this. The problem is linked to the WaitForAccessRuleDeleted method. When the access rule is deleted, it gets removed almost immediately from the list, so when I call GetAccessRules, it shows the rule has been removed. But from the ActivityList, you can still see the operation is still running.
The WaitForAccessRuleDeleted wrongly exits from the wait when it no longer sees the AccessRule in the returned list. I have changed the method to wait until there is no activity from the activity log being returned as well.
Not too sure if the other libraries will have this problem.
This was also causing the defer of the DeleteAccessRule and DeleteServiceInstance for fail.

Issue is now fixed, and it should work properly now.